lib/tree_sequence.c

```diff
@@ -986,6 +1001,300 @@ tree_sequence_simplify(tree_sequence_t *self, node_id_t *samples, size_t num_sam
+            {
+                mutation_t *mut = &site->mutations[j];
+                node_map_t *node_map_entry = node_map + node_map__count;
+
+                node_map_entry->site_index = j;
+                node_map_entry->node_id = mut->node;
+                node_map_entry->mutation_id = mut->id;
+                node_map_entry->branch_group_checked = false;
+                node_map__count++;
+            }
+
+            // Sort the node map by node id (and by site index as the sub-key),
+            // which makes finding/handling branch groups (see below) easier
+            qsort(node_map, node_map__count, sizeof(node_map_t),
+                    &node_map_compare_node_id);
+
+            // Now we look for "branch groups", sets of mutations that are on
+            // the same branch (i.e. have the same node id), and verify that
+            // each branch group forms a single-file chain.  Since we have
+            // sorted the node map by node id, branch groups will now be
+            // adjacent in the node map.  If the branch group is well-formed,
+            // each entry is the parent of the entry directly below it, so we
+            // can just scan the branch group in order and check parentage.
+            node_id_t previous_node_id = node_map[0].node_id;
+
+            for (table_size_t j = 1; j < node_map__count; ++j)
+            {
+                node_id_t current_node_id = node_map[j].node_id;
+
+                if (current_node_id == previous_node_id)
+                {
+                    // a branch group starts at j - 1 (previous_node_id); find its extent
+                    table_size_t first_in_group = j - 1, last_in_group = j;
+
+                    // the chain end needs to be checked below
+                    node_map[first_in_group].branch_group_checked = false;
+                    node_map[last_in_group].branch_group_checked = true;
+
+                    while ((last_in_group + 1 < node_map__count)
+                            && (node_map[last_in_group + 1].node_id == previous_node_id))
+                    {
+                        ++last_in_group;
+                        node_map[last_in_group].branch_group_checked = true;
+                    }
+
+                    // check that the branch group forms a linear chain
+                    for (table_size_t group_index = first_in_group + 1;
+                            group_index <= last_in_group; ++group_index)
+                    {
+                        mutation_id_t putative_parent_mut_id = node_map[group_index - 1].mutation_id;
+                        table_size_t mut_site_index = node_map[group_index].site_index;
+                        mutation_id_t actual_parent_mut_id = site->mutations[mut_site_index].parent;
+
+                        if (fix_missing_parents && (actual_parent_mut_id == MSP_NULL_MUTATION)) {
+                            site->mutations[mut_site_index].parent = putative_parent_mut_id;
```

Comment:
IIUC, this bit doesn't actually record the parents in the tree sequence, only in the (temporary) mutations that are kept by that site: it should be
```
self->mutations.parent[mutation_table_index] = putative_parent_mut_id;
```
where `mutation_table_index` also needs to be defined to keep track of which mutation in the overall tree sequence we're looking at.

Is this right, @jeromekelleher? And, if we dynamically update the `tree_sequence.mutations.parent` array, is there anything else in the tree sequence that needs updating afterwards?  I think not...